doc/TclEvent.3
'\"
'\" Copyright (c) 1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
.so man.macros
.TH TclEvent 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.TclEvent -- Subclass from this abstract base class to
create event handlers.

.SH METHODS
.sp
abstract int \fBprocessEvent\fR(int \fIflags\fR)
.sp
synchronized final void \fBsync\fR()

.SH ARGUMENTS
.AS TclObject "table[]" .
.AP int flags
Same as \fIflags\fR passed to the \fBdoOneEvent\fR method of the
\fBNotifier\fR class.
.BE

.SH DESCRIPTION

The TclEvent class is used to schedule events to be processed in a
Jacl Notifier. Usually, an event-driven Jacl application consists of
one "Notifier Thread" and one or more "Event Detection Threads". When
an Event Detection Thread detects an event, it puts an instance of
TclEvent on the queue of the Notifier. Then, when the Notifier Thread
is ready to process events (i.e., when the \fBdoOneEvent\fR method of
the \fBNotifier\fR class is invoked), it invokes the
\fBprocessEvent\fR method of the TclEvent instances in the order they
appear in the Notifier's event queue.

.TP
\fBprocessEvent\fR
When \fBprocessEvent\fR is called, it should determine whether the
event can be processed right away (possibly based on the settings of
the \fIflags\fR parameter). If the event cannot be processed yet, this
method returns 0, in which case the Notifier will leave the event in
the event queue and defer its processing to some later
time. Otherwise, this method should process the event and return 1, in
which case the Notifier will remove the event from the event queue.

.TP
\fBsync\fR
Usually, \fBTclEvent\fR instances are processed asynchronously. If
synchronous processing is desired, an Event Detection Thread can call
the \fBsync\fR method, which waits until the event is processed by the
Notifier. The TclEvent must be in the event queue before its
\fBsync\fR method can be invoked, otherwise a \fBTclRuntimeError\fR
will be generated.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_EventProc, Tcl_ServiceAll

.SH "SEE ALSO"
Notifier, Interp, TclRuntimeError, IdleHandler, TimerHandler

.SH KEYWORDS
event, event queue, thread, event sources, event detection


ChangeLog
2019-05-04  Fabrice Ducos <fabrice.ducos@gmail.com>

        * Support of latest versions of Tcl (at least Tcl 8.5
        and Tcl 8.6) and of modern JDK (10+);
        support of JDK 4- abandoned (use of auto-boxing);
        support of OSX systems added;
        details below

        * configuration (tcljava.m4, configure.in),
        and compilation (Makefile.in) templates updated for
        JDK 10+ (javah not available anymore).
        javah is still used when it is available (JDK 9-).
        * minor update of jtclsh.in (for OSX dynamic libraries)
        * many obsolescent constructs (mostly old-style boxing)
        removed in *.java under src/jacl/tcl/lang,
        src/tclblend/tcl/lang and src/tcljava/tcl/lang,
        in order to silence many warnings.
        (effect: support of JDK 4- abandoned)
        * src/native/javaInterp.c: 2010 implementation
        of Java_tcl_lang_Interp_initName() crashed on the
        example given at docs/website/getstart.html,
        because calling Tcl_GetNameOfExecutable
        before Tcl_FindExecutable gives rise to undefined
        behaviour, at least in latest versions of Tcl
        (crashes observed on several systems, at least with
        Tcl 8.6).
        Java_tcl_lang_Interp_initName() is now simplified and
        always calls Tcl_FindExecutable first.
        * all the occurences of interp->result have been
        replaced with Tcl_GetStringResult(interp).
        Accessing interp->result directly was deprecated in
        latest versions of the Tcl C API, and removed from
        Tcl 8.6 (another solution was to #define USE_INTERP_RESULT,
        but this would have been an easy-going, quick and dirty way)
        ref. https://www.tcl.tk/man/tcl/TclLib/Interp.htm

2019-05-01  Fabrice Ducos <fabrice.ducos@gmail.com>

        * tcljava resurrected
        * migration from a CVS to a GIT repository
        (with reposurgeon)

2010-02-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* tests/tcl/regexp.test: Add test case for
        escaped - character in regexp command.
        [Bug 2012389]

2010-02-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceAll):
	Fixup regsub -all logic so 5 more tests pass.
	The replaceAll() logic is broken because it depends
	on matching substrings instead of using an index
	into the larger string and Java's regexp library
	uses this to calculate line starts in multiline mode.

2010-02-18  Mo DeJong  <mdejong@users.sourceforge.net>

        Incorporate fix/tests from Tcl head on 2010-02-11,
        aka was Tcl [Bug 2826551, Patch 2948425].

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
        Fixup increment index on zero length match issue.
	* tests/tcl/regexp.test: Back port of new regexp tests,
        some tests marked with {endIndex} due to lack of impl.

2010-02-11  Mo DeJong  <mdejong@users.sourceforge.net>

	Patch 2946118

	* src/jacl/tcl/lang/ListCmd.java: Minor style reformat.
	* src/jacl/tcl/lang/Util.java (scanElement, convertElement):
	Remove crusty old code that treated backslash zero
	as the end of the line. That logic was removed from
	C Tcl long ago.
	* tests/inprogress/EventAdaptor.test: Fixup tests
	that depended on backslash zero quirks.
	* tests/tcl/list.test: Add test cases for backslash zero
	as argument to list command.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (match, _match,
        replaceFirst, replaceAll): Rewrite -start offset
        logic to correctly deal with odd edge cases.
	* tests/tcl/regexp.test: Add additional test cases
        that cover -start index edge cases.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Rewrite argument processing logic to avoid using
        IndexOutOfBoundsException to deal with an incorrect
        number of arguments.

2009-10-03  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fix regsub -all crash because offset was not getting
        advanced by 1 on a match of length zero.
        Fixup -start argument support with and without -all
        in the case where the offset was larger than the
        length of the string.
	* tests/tcl/regexp.test: Add additional test cases.
        Enable regsub test cases that were getting caught
        in an infinite loop.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        failing tests related to optional subgroups.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Minor paren fix, no function changes.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Rework
        logic related to match result groups.
	* tests/tcl/regexp.test: Add missing test cases for
        non-capturing groups and -inline results.

2009-09-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup use of MULTILINE flag.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for -linestop and -lineanchor arguments to regexp.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):  Fixup
        support for -linestop and -lineanchor arguments to regsub.
	* tests/tcl/regexp.test: Add missing test cases for
        -linestop and -lineanchor arguments.

2009-09-22  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup support for the empty string as the string
        argument to regsub by removing the MULTILINE flag when
        an empty string is detected.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for empty string as string argument.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Remove comments.
	* tests/tcl/regexp.test: Add additional test cases

tcljava/src/tcljava/tcl/lang/TclEvent.java
/*
 * TclEvent.java --
 *
 *	Abstract class for describing an event in the Tcl notifier
 *	API.
 *
 * Copyright (c) 1997 Cornell University.
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: TclEvent.java,v 1.2 2000/01/25 03:42:26 mo Exp $
 *
 */

package tcl.lang;

/*
 * This is an abstract class that describes an event in the Jacl
 * implementation of the notifier. It contains package protected
 * fields and methods that are accessed by the Jacl notifier. Tcl Blend
 * needs a different implementation of the TclEvent base class.
 *
 * The only public methods in this class are processEvent() and
 * sync(). These methods must appear in both the Jacl and Tcl Blend versions
 * of this class.
 */

public abstract class TclEvent {

/*
 * The notifier in which this event is queued.
 */

Notifier notifier = null;

/*
 * This flag is true if sync() has been called on this object.
 */

boolean needsNotify = false;

/*
 * True if this event is current being processing. This flag provents
 * an event to be processed twice when the event loop is entered
 * recursively.
 */

boolean isProcessing = false;

/*
 * True if this event has been processed.
 */

boolean isProcessed = false;

/*
 * Links to the next event in the event queue.
 */

TclEvent next;


/*
 *----------------------------------------------------------------------
 *
 * processEvent --
 *
 *	Process the event. Override this method to implement new types
 *	of events.
 *	
 *	Note: this method is called by the primary thread of the
 *	notifier.
 *
 * Results:	
 *	1 means the event has been processed and can be removed from
 *	the event queue. 0 means the event should be deferred for
 *	processing later.
 *
 * Side effects:
 *	This method may have arbitrary side effects while processing
 *	the event.
 *
 *----------------------------------------------------------------------
 */

public abstract int
processEvent(
    int flags);			// Same as flags passed to Notifier.doOneEvent.


/*
 *----------------------------------------------------------------------
 *
 * sync --
 *
 *	Wait until the event is processed.
 *
 * Results:	
 *	None.
 *
 * Side effects:
 *	Arbitrary things may happen while this thread is waiting.
 *
 *----------------------------------------------------------------------
 */

public final void
sync()
{
    if (notifier == null) {
	throw new TclRuntimeError(
		"TclEvent is not queued when sync() is called");
    }

    if (Thread.currentThread() == notifier.primaryThread) {
	while (!isProcessed) {
	    notifier.serviceEvent(0);
	}
    } else {
	synchronized (this) {
	    needsNotify = true;
	    while (!isProcessed) {
	        try {
		    wait(0);
	        } catch (InterruptedException e) {
		    // Another thread has sent us an "interrupt"
		    // signal. We ignore it and continue waiting until
		    // the event is processed.

		    continue;
	        }
	    }
	}
    }
}

} // end TclEvent

